### ISSUE TITLE
Add further swipe actions

### ISSUE THREAD

**Issue by Glitchy-Tozier** [florisboard/florisboard#215]
Additional ideas for the left/right swipe across the keyboard:

- Switch between keyboards. This could be used to cycle between the normal keyboard, symbol keyboard 1&2 and the selection keyboard. (Maybe even emojies.)
In that case, If both left-swipe and right-swipe are used, they should cycle through the keyboards in the opposite order of each other.
- Change between the top-bar functionalities (The copy&past bar, the undo-bar, and, eventually, the word suggestions)

- Undo
- Redo

Of course all those Swiping-Actions should then be available in every keyboard.

**Comment by patrickgold** [comment]
Thanks for the suggestion!

> Switch between keyboards. This could be used to cycle between the normal keyboard, symbol keyboard 1&2 and the selection keyboard

This is a nice idea I haven't thought about yet. Though I think the selection keyboard should not be included in the "swipe cycle".

> (Maybe even emojies.)

Nope, because this interferes with the scroll view in the emoji media input.

> In that case, If both left-swipe and right-swipe are used, they should cycle through the keyboards in the opposite order of each other.

That's how I would implement it too.

> Change between the top-bar functionalities (The copy&past bar, the undo-bar, and, eventually, the word suggestions)

This idea sounds pretty cool, especially if I add separate left right swipes for the Smartbar. This would also be one solution for the often stated request to be able to quick switch between the suggestion strip and the clipboard tools. The "undo-bar" (which I call the quick actions) though should not be included within this swipe cycle, as there's the toggle button on the left, which is almost always available (except in the selection keyboard and when entering passwords).

Of course I will add all the swipe actions to any swipe preference (except key delete), the Smartbar swipes will get the toll switching as the default action. Will add it to my todo list!

**Comment by Glitchy-Tozier** [comment]
> Though I think the selection keyboard should not be included in the "swipe cycle".

I think at the very least it should be a different option then.
Reason: The selection-keyboard is the hardest to reach out of all of them. It requires more button presses / the most movement out of all the keyboards.


Oh, I never thought about this... Is the way the top bars are structured this?

Quick Actions / (Copy&Paste / Word Predictions)

Or like this?

Quick Actions / Copy&Paste / Word Predictions


Are they all on the same level or is it "`Quick Actions` and `Everything Else`"?

**Comment by patrickgold** [comment]
>  Reason: The selection-keyboard is the hardest to reach out of all of them. It requires more button presses / the most movement out of all the keyboards.

That's actually a good point, the editing layout isn't quickly accessible. Still I don't like it to be included in the swipe cycle. I will definitely though have to find some sort of solution for this.

> Oh, I never thought about this... Is the way the top bars are structured this?

On a technical level the structure is `Quick Actions / Copy&Paste / Word Predictions` (a flat container of all Smartbar features)

On a user's perspective level (at least how I wanted it to be): `Quick Actions / (Copy&Paste / Word Predictions)`, so quick actions and then everything else. The quick action toggle button is there to expand the quick action menu and should show that you just expanded a overlay menu which is above the current Smartbar feature. If I include it in the Smartbar swipe cycle, it seems that it is just another container, which I do not want it to be.

**Comment by Glitchy-Tozier** [comment]
I guess to "show" that it is an overlay, a short & quick sliding-over-animation could do the trick (left to right, probably).
(Closing quick-actionsshould subsequently be a slide-away-animation)

In general, that sounds like a pretty good idea.

_____________
How many different Smartbars are currently planned? (not counting quick actions)

**Comment by patrickgold** [comment]
> I guess to "show" that it is an overlay, a short & quick sliding-over-animation could do the trick (left to right, probably).
(Closing quick-actionsshould subsequently be a slide-away-animation)

Animations are a great idea, but have been overly avoided by me because I have never designed one yet. Time to learn something new for me :)

> How many different Smartbars are currently planned? (not counting quick actions)

There's currently no active additional toolbar row planned to be added. I plan though to add passive toolbars (especially for better clipboard suggestion, emoji suggestion, emoji quick search etc.) which only activate in specific circumstances and which look similar to the suggestion strip (but are technically a separate toolbar). So they will never be directly reachable by a button or a swipe action, only if some event or state triggers it.

**Comment by Glitchy-Tozier** [comment]
> Animations are a great idea, but have been overly avoided by me because I have never designed one yet. Time to learn something new for me :)

Exciting!!!  :)

> There's currently no active additional toolbar row planned to be added. I plan though to add passive toolbars (especially for better clipboard suggestion, emoji suggestion, emoji quick search etc.) which only activate in specific circumstances and which look similar to the suggestion strip (but are technically a separate toolbar). So they will never be directly reachable by a button or a swipe action, only if some event or state triggers it.

I didn't mean to ask _"how many will there be stacked on top of each other?"_  but rather _"Will there be multiple ones you can switch between without going to a different board or changing the settings?"_
(Again, not counting the quick actions)

**Comment by patrickgold** [comment]
> I didn't mean to ask "how many will there be stacked on top of each other?" but rather "Will there be multiple ones you can switch between without going to a different board or changing the settings?"
(Again, not counting the quick actions)

No, I currently do not have plans to have entire different Smartbars alongside each other.

**Comment by Glitchy-Tozier** [comment]
Then how many sub-bars are planned?  😅

(=How many (and which ones) can you switch between without changing the setting inbetween?)

**Comment by patrickgold** [comment]
Ok I think we talk past one another here, so I will try to explain the structure of FlorisBoard's text input UI, which is as follows:

```xml
<TextInputUi>
    <Smartbar ... />
    <KeyboardViewContainer ... />
</TextInpuUi>
```

From the above structure you can see there's exactly one Smartbar (and there will always be exactly one Smartbar in the future).
This is the structure of the Smartbar:

```xml
<Smartbar>
    <QuickActionToggleButton ... /><!-- can hide itself if needed -->
    <!-- Main area can only show 1 of its children at once. A child of this area is called a "row" or a "sub-bar". -->
    <MainArea>
        <SuggestionStrip swipecycle=yes ... />
        <SuggestionStripVariant1Clipboard ... />
        <SuggestionStripVariant2EmojiSuggestion ... /><!-- not actually implemented yet, but planned -->
        <SuggestionStripVariant3EmojiQuickSearch ... /><!-- not actually implemented yet, but planned -->
        <CliboardCursorRow swipecycle=yes ... />
        <NumberRowSmartbarEdition ... /><!-- Only shown for password fields & when "real" number row hidden -->
        <!-- Room to add more rows aka "sub-bars" here -->
    </MainArea>
    <PrivateModeButton ... /><!-- Can hide itself when not in private mode -->
</Smartbar>
```

From the above structure you can see that there are actually a lot of different sub-bars within the Smartbar. At the current state, you **cannot** switch from one sub-bar to the next without going to the Settings. With the gesture left/right idea however you will be able to swipe between all sub-bars which are tagged with `swipecycle=yes`. Atm I have no sub-bars planned which are included in the swipe cycle except clipboard row and suggestion strip. Of course if I or someone else has an idea for a new one, I am fully open to add one. When the Smartbar gesture left/right is implemented, you will be able to cycle through the clipboard cursor row and the suggestion strip **without** going to the Settings (unless you set these swipe actions to something else). If you fully disable suggestions, swipe left/right will then do basically nothing, as there's no more sub-bar to switch to.

I hope this can explain the structure and terminology of the Smartbar better to you and this clears up some confusion :)

**Comment by Glitchy-Tozier** [comment]
I see, that helps. Thank you for answering all my questions!  :)

**Comment by patrickgold** [comment]
In above PR several new gestures have been added:
- Undo
- Redo
- Cycle to previous keyboard mode (cycle is between CHARACTERS, SYMBOLS, SYMBOLS2 and NUMERIC_ADVANCED currently). The selection keyboard is its own thing and thus currently not included in the cycle.
- Cycle to next keyboard mode (the same thing but backwards).

I've omitted the Smartbar gestures, as I've introduced scrolling candidates in v0.3.10, which completely interfere with adding swipe left/right to the Smartbar. Nevertheless the Smartbar will get some options for switching between the modes in the future, it just won't be through gestures. Will be released in v0.3.11-beta01.
